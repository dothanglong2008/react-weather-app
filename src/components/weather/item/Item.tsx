import { useState } from "react";
import styles from "./Item.module.scss";
import { IoIosArrowUp, IoIosArrowDown } from "react-icons/io";

interface HourlyItemProps {
  time: string;
  icon: string;
  temp: number;
  humidity: number;
  chanceOfRain: number;
  extraTable: Array<{ field: string; value: string | number }>;
}

function Item({
  time,
  icon,
  temp,
  humidity,
  chanceOfRain,
  extraTable,
}: HourlyItemProps) {
  const [extra, setExtra] = useState(false);
  return (
    <>
      <li
        className={styles.itemContainer}
        onClick={() => {
          setExtra((extra) => !extra);
        }}
      >
        <table className={styles.primary}>
          <tbody className={styles.primaryBody}>
            <tr className={styles.primaryRow}>
              <td>
                <strong>{time}</strong>
              </td>
              <td className={styles.condition}>
                <span className={styles.icon}>
                  <img src={icon} alt="weather icon" />
                </span>
                <strong>{temp}&#8451;</strong>
              </td>
              <td className={styles.humidity}>{humidity}</td>
              <td className={styles.rain}>{chanceOfRain}%</td>
              <td className={styles.arrow}>
                {extra ? <IoIosArrowUp /> : <IoIosArrowDown />}
              </td>
            </tr>
          </tbody>
        </table>
        {extra && (
          <table className={styles.extra}>
            <tbody className={styles.extraBody}>
              {extraTable.map((data) => (
                <tr key={data.field} className={styles.extraRow}>
                  <td>{data.field}</td>
                  <td>{data.value}</td>
                </tr>
              ))}
            </tbody>
          </table>
        )}
      </li>
    </>
  );
}

export default Item;
