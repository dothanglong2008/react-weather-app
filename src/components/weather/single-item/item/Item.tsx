import styles from "./Item.module.scss";

interface ItemProps {
  last_updated?: string;
  last_updated_epoch?: number;
  temp_c?: number;
  temp_f?: number;
  feelslike_c?: number;
  feelslike_f?: number;
  condition?: {
    text?: string;
    icon?: string;
    code?: number;
  };
  wind_mph?: number;
  wind_kph?: number;
  wind_degree?: number;
  wind_dir?: string;
  pressure_mb?: number;
  pressure_in?: number;
  precip_mm?: number;
  precip_in?: number;
  humidity?: number;
  cloud?: number;
  is_day?: number;
  uv?: number;
  gust_mph?: number;
  gust_kph?: number;
}

function Item({
  last_updated,
  condition,
  temp_c,
  temp_f,
  feelslike_c,
  uv,
  precip_mm,
}: ItemProps) {
  return (
    <article className={styles.item}>
      <h1 className={styles.item}>CURRENT WEATHER</h1>
      <p className={styles.item}>{last_updated}</p>
      <h2 className={styles.item}>
        <span className={styles.item}>{condition?.icon}</span>
        <span className={styles.item}>{temp_c}</span>
        <span className={styles.item}>&#8451;</span>
        <span className={styles.item}>RealFeel</span>
        <span className={styles.item}>{feelslike_c}</span>
      </h2>
      <p className={styles.item}>
        <span className={styles.item}>{temp_f}</span>
        <span className={styles.item}>&#8457;</span>
      </p>
      <p className={styles.item}>{condition?.text}</p>
      <div className={styles.item}>
        <table className={styles.item}>
          <tbody className={styles.item}>
            <tr className={styles.item}>
              <th className={styles.item}>Max UV Index</th>
              <td className={styles.item}>{uv}</td>
              <th className={styles.item}>Precipitation</th>
              <td className={styles.item}>{precip_mm}</td>
            </tr>
            <tr className={styles.item}>
              <th className={styles.item}>Wind</th>
              <td className={styles.item}>{uv}</td>
              <th className={styles.item}>Rain</th>
              <td className={styles.item}>{precip_mm}</td>
            </tr>
            <tr className={styles.item}>
              <th className={styles.item}>Wind Gusts</th>
              <td className={styles.item}>{uv}</td>
              <th className={styles.item}>Hours of Precipitation</th>
              <td className={styles.item}>{precip_mm}</td>
            </tr>
          </tbody>
        </table>
      </div>
    </article>
  );
}

export default Item;
