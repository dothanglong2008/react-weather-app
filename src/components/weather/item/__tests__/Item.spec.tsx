import Item from "../Item";
import renderer from "react-test-renderer";

describe("Test Item component", () => {
  it("renders Item correctly", () => {
    const table = [
      { field: "RealFeel", value: 1 },
      { field: "Wind", value: 2 },
      { field: "Wind Gust", value: 3 },
      { field: "Humidity", value: 4 },
      { field: "Dew Point", value: 5 },
      { field: "UV", value: 1 },
    ];

    const tree = renderer
      .create(
        <Item
          key={1}
          time={"02:00"}
          icon={"icon"}
          temp={37}
          humidity={2}
          chanceOfRain={1}
          extraTable={table}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
