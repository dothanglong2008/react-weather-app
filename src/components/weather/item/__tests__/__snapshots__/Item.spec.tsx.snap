// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Item component renders Item correctly 1`] = `
<li
  className="itemContainer"
  onClick={[Function]}
>
  <table
    className="primary"
  >
    <tbody
      className="primaryBody"
    >
      <tr
        className="primaryRow"
      >
        <td>
          <strong>
            02:00
          </strong>
        </td>
        <td
          className="condition"
        >
          <span
            className="icon"
          >
            <img
              alt="weather icon"
              src="icon"
            />
          </span>
          <strong>
            37
            â„ƒ
          </strong>
        </td>
        <td
          className="humidity"
        >
          2
        </td>
        <td
          className="rain"
        >
          1
          %
        </td>
        <td
          className="arrow"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
            />
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
</li>
`;
